specs/interop/messaging.md

```diff
@@ -136,13 +149,121 @@ This feature could be added in a backwards compatible way by adding a new functi
 One possible way to handle explicit denial of service attacks is to utilize identity
 in iterated games such that the block builder can ban identities that submit malicious transactions.
 
-### Message Expiry
+### Message Expiry Invariant
+
+Note: Message Expiry as property of the protocol is in active discussion.
+It helps set a strict bound on total messaging activity to support, but also limits use-cases.
+This trade-off is in review. This invariant may be ignored in initial interop testnets.
+
+The expiry invariant invalidates inclusion of any executing message with
+`id.timestamp + EXPIRY_TIME < executing_block.timestamp` where:
+
+- `id` is the [`Identifier`] encoded in the executing message, matching the block attributes of the initiating message.
+- `executing_block` is the block where the executing message was included in.
+- `EXPIRY_TIME = 180 * 24 * 60 * 60 = 15552000` seconds, i.e. 180 days.
+
+## Message Graph
+
+The dependencies of messages can be modeled as a directed graph:
+
+- **vertex**: a block
+- **edge**: a dependency:
+  - parent-block (source) to block (target)
+  - message relay, from initiation (source) to execution (target)
+
+If the source of an edge is invalidated, the target is invalidated.
+
+### Invalid messages
+
+A message is said to be "invalid" when the executing message does not have a valid dependency.
+
+Dependencies are invalid when:
+
+- The dependency is unknown.
+- The dependency is known but not part of the canonical chain.
+- The dependency is known but does not match all message attributes ([`Identifier`] and payload).
+
+#### Block reorgs
+
+The [`Identifier`] used by an executing message does not cryptographically
+commit to the block it may reference.
+
+Messages may be executed before the block that initiates them is sealed.
+
+When tracking message dependencies, edges are maintained for *all* identified source blocks.
+
+Reorgs are resolved by filtering the view of the solver to only canonical blocks.
+If the source block is not canonical, the dependency is invalid.
+
+The canonical L2 block at the identified block-height is the source of truth.
+
+#### Block Recommits
+
+Blocks may be partially built, or sealed but not published, and then rebuilt.
```

Comment:
There is a lot of hidden context in this paragraph, it could be made more approachable